### ISSUE TITLE
[BUG][Fuzzer][wrong-code] Writing a fragment through a `T.serial` loop: silently wrong results (concurrent threads race-write each cell) instead of correct output or an error

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2396]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11+cuda.gita4399e4c

### System information

NVIDIA A10G (sm_86), CUDA 12.8 (nvcc V12.8.93), PyTorch 2.8.0, Python 3.11. `execution_backend="cython"`.

### Problem description

Writing a fragment to global memory through a `T.serial` loop compiles and runs with no error, but
produces silently wrong output. With a 16×16 `float32` fragment and a distinct-value input
(`arange(0..255)`), 250 of 256 output cells are wrong (max abs error 252.0); every output row collapses
to `[0, 2, 0, 2, …]`, deterministically. The identical kernel with `T.Parallel(M, N)` in place of the
`T.serial` loop is exactly correct (max error 0.0).

In the generated CUDA, every thread runs the full loop, the fragment is read with an index that has no
`threadIdx.x` term (`Af[j & 1]`), and the store is gated by `if (threadIdx.x <= …)` — so multiple threads
write the same global cell with different values:
```cuda
float Af[2];
*(float2*)(Af + 0) = *(float2*)(A + (threadIdx.x * 2));
for (int i = 0; i < 16; ++i)
  for (int j = 0; j < 16; ++j)
    if (((int)threadIdx.x) <= ((i * 8) + (j >> 1)))
      C[((i * 16) + j)] = (Af[(j & 1)] * 2.0f);
```

### Reproducible example code

```python
import torch, tilelang, tilelang.language as T
M = N = 16

@T.prim_func
def kernel(A: T.Tensor((M, N), "float32"), C: T.Tensor((M, N), "float32")):
    with T.Kernel(1, 1, threads=128) as (bx, by):
        Af = T.alloc_fragment((M, N), "float32")
        T.copy(A[0, 0], Af)
        for i in T.serial(M):
            for j in T.serial(N):
                C[i, j] = Af[i, j] * 2.0

kfn = tilelang.compile(kernel, target="cuda")
a = torch.arange(M * N, dtype=torch.float32, device="cuda").reshape(M, N)   # distinct values
c = torch.empty(M, N, dtype=torch.float32, device="cuda")
kfn(a, c); torch.cuda.synchronize()
diff = (c - a * 2.0).abs()
print("max err:", diff.max().item())                            # 252.0
print("wrong cells:", (diff > 1e-4).float().mean().item())      # 0.977
print("got [0,:8]:", c[0, :8].tolist())                         # [0, 2, 0, 2, 0, 2, 0, 2]
print("want[0,:8]:", (a * 2.0)[0, :8].tolist())                 # [0, 2, 4, 6, 8, 10, 12, 14]
```
Replacing the serial loop with `for i, j in T.Parallel(M, N): C[i, j] = Af[i, j] * 2.0` gives max
error 0.0. (Use a distinct-value input — a constant input masks the bug, since colliding threads then
all hold the same value.)

### Traceback

```pytb
None — the kernel compiles and runs cleanly; the output is silently wrong.
```

### Expected behavior

Either produce the correct result (as the `T.Parallel` form does), or reject the kernel at compile time
with a clear message — not silently wrong output.

### Additional context

`tl.VerifyParallelLoop` (the data-race check, on by default) runs on this kernel but does not flag it: it
only tracks `ForKind::kParallel` loop variables ([`verify_parallel_loop.cc:30`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/transform/verify_parallel_loop.cc#L30)),
so a store inside a `T.serial` loop is checked with no thread variables and reports no race.

Other manifestations of a `T.serial` loop over a fragment: an NVRTC "identifier undefined" crash when
the serial-written fragment is then `T.copy`'d to global (filed alongside this issue), and a `T.serial`
reduce over a GEMM-output fragment raising an unhelpful internal error (#2393).
